src/main/java/com/aws/greengrass/mqtt/bridge/MQTTBridge.java

```diff
@@ -119,60 +123,66 @@ protected MQTTBridge(Topics topics, TopicMapping topicMapping, MessageBridge mes
 
     @Override
     public void install() {
-        configurationChangeHandler.listen();
-        messageBridge = new MessageBridge(this.topicMapping, bridgeConfig.get().getMqtt5RouteOptionsForSource(
-                TopicMapping.TopicType.LocalMqtt));
+        synchronized (lifecycleLock) {
+            configurationChangeHandler.listen();
+            messageBridge = new MessageBridge(this.topicMapping, bridgeConfig.get().getMqtt5RouteOptionsForSource(
+                    TopicMapping.TopicType.LocalMqtt));
+        }
     }
 
     @Override
     public void startup() {
-        try {
-            mqttClientKeyStore.init();
-        } catch (KeyStoreException | CertificateGenerationException e) {
-            serviceErrored(e);
-            return;
-        }
+        synchronized (lifecycleLock) {
+            try {
+                mqttClientKeyStore.init();
+            } catch (KeyStoreException | CertificateGenerationException e) {
+                serviceErrored(e);
+                return;
+            }
 
-        certificateAuthorityChangeHandler.start();
+            certificateAuthorityChangeHandler.start();
 
-        try {
-            localMqttClient = localMqttClientFactory.createLocalMqttClient();
-            localMqttClient.start();
-            messageBridge.addOrReplaceMessageClientAndUpdateSubscriptions(
-                    TopicMapping.TopicType.LocalMqtt, localMqttClient);
+            try {
+                localMqttClient = localMqttClientFactory.createLocalMqttClient();
+                localMqttClient.start();
+                messageBridge.addOrReplaceMessageClientAndUpdateSubscriptions(
+                        TopicMapping.TopicType.LocalMqtt, localMqttClient);
 
-            pubSubClient.start();
-            messageBridge.addOrReplaceMessageClientAndUpdateSubscriptions(
-                    TopicMapping.TopicType.Pubsub, pubSubClient);
+                pubSubClient.start();
+                messageBridge.addOrReplaceMessageClientAndUpdateSubscriptions(
+                        TopicMapping.TopicType.Pubsub, pubSubClient);
 
-            ioTCoreClient.start();
-            messageBridge.addOrReplaceMessageClientAndUpdateSubscriptions(
-                    TopicMapping.TopicType.IotCore, ioTCoreClient);
+                ioTCoreClient.start();
+                messageBridge.addOrReplaceMessageClientAndUpdateSubscriptions(
+                        TopicMapping.TopicType.IotCore, ioTCoreClient);
 
-            reportState(State.RUNNING);
-        } catch (MessageClientException e) {
-            serviceErrored(e);
+                reportState(State.RUNNING);
+            } catch (MessageClientException e) {
+                serviceErrored(e);
+            }
         }
     }
 
     @Override
     public void shutdown() {
-        certificateAuthorityChangeHandler.stop();
```

Comment:
seems like the main problem is that shutdown either isn't invoked or it isn't actually stopping the client fully thus allowing it to keep connecting.

Reply:
yeah, i've noticed in debugging that the client can be null during shutdown (even if client variable is volatile)

Reply:
so if the timing is unfortunate enough, shutdown happens just before client is created during startup, creating the leak

Reply:
found that checking for interrupt before creating/starting the client works too, but there's still a race involved technically

Reply:
how can that happen though? startup would be interrupted, so it should then shutdown

Reply:
OK, so startup is interrupted, but that isn't resulting in the client _not_ trying to connect?

Reply:
was thinking about this scenario
```
startup() {
   if (interrupted) {
       return;
   }
   // shutdown() could execute theoretically before the next line, causing a leak?
   client = createMQTTClient()
}
```

Reply:
yes, but creating the client should fail (possibly after attempting to connect) _or_ the client should not be re-created when starting again

Reply:
ah i see, will check that